### PR TITLE
fix: [CDS-83890]: Implement shared helm commands

### PR DESCRIPTION
**Pull request by abosii-harness**
## Describe your changes
As part of these changes we move the helm commands in the next 3 categories:
- [x] Shared commands - these commands execution is shared between all threads that are currently require to run a specific command. This means that if there are 2 threads that are running the same command at the same time then instead of creating 2 processes we will create instead only one and share the result with these threads. 
- [x] Local commands - these commands are executed by the each thread
- [x] Thread Pool commands - similar to local commands, but the actual execution is done by the thread pool used by shared commands. This allows to limit the number of parallel running processes

### How commands are categorized
Currently there are 3 criterias:
1. If command is parsing repository index.yaml and the process response is independent on the execution that is running it then this will be a shared command. For example, helm repo add will produce the same result independent of the which deployment is running it.
2. If command is parsing repository index.yaml but the process response depends on the execution that is running it then this will be a thread pool command. For example, helm pull will download the helm chart to the configured working directory defined by each deployment. 
3. Everything else are local commands

### Helm Commands Table
| Type | Commands | Why? |
| --- | --- | --- |
| Shared | helm repo add, helm repo update, helm search repo | All these commands are parsing index.yaml and will always produce the same result and doesn't depend on the current working directory |
| Thread Pool |  helm pull | Helm pull will parse the index.yaml as well but it's dependent on the working directory (it does store the fetched helm chart) |
| Local | everything else | These commands are not expected to end up in parsing index.yaml or they haven't been spoted to cause any cpu spikes, for example helm version | 

### How this solves customer issue
While this solution can't fully solve the high CPU usage it does allow to reduce the number of concurrently running helm repo add and helm repo update as well provide a way to control how many helm commands that can cause CPU spike can run in parallel. 

### How customer can limit the number of parallel helm commands 
Customer can export env variable `HELM_CLI_MAX_POOL_SIZE` that will configure the thread pool max threads count

### How shared commands are working
We're going to keep track of currently running processes based on `processKey`. In this case `processKey` matches the executed command since all commands that are running on specific repository are defined by `--repository-config` argument this approach is good enough to identify a command that is running on specific repository. 
When a deployment wants to run a command, we will check if there is a similar command already running and return the future of the currently running command if does exist or queue a new one if doesn't. Each command reference will increase current reference count of the running command and decrease once either the thread waiting on command has been interrupted or completed. Once the reference count reach 0, the thread that was the latest that referenced that command will remove the command from running and interrupt the execution if it's the case. 

## Test cases
- [x] Run 10-15 deployments in parallel using a repository with a small index.yaml
- [x] Run 10-15 deployments in parallel using a repository with index.yaml of size of 25 Mb
- [x] Run 10-15 deployments in parallel using a repository with index.yaml of size of 75 Mb
- [x] Run 10-15 deployments in parallel using a repository with index.yaml of size of 150 Mb
- [x] Run 10-15 deployments in parallel using multiple repository (small, 25 mb, 75mb and 150 mb)
- [x] In case if delegate crashes adjust the `HELM_CLI_MAX_POOL_SIZE`
- [x] Run K8s Deployment with Helm
- [x] Run Native Helm Deployment
- [x] Test the GCS, S3 helm chart
- [x] Test OCI helm chart

### Observations
1. In case of repo size 75 mb and 150 mb the pool size that doesn't crashes delegate of 2 cpu and 4096 mb is 5 threads
2. In case of combination of 25 mb, 75 mb and 150 mb deployments, even with the pool size of 5 threads, for delegate of  2 cpu and 4096 mb, sometimes delegate still crashes
3. The main reason behind these crashes are `helm pull` commands.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55409)

### CODE REVIEW

**Comment by abosii-harness** [comment]
trigger publish-delegate

**Comment by abosii-harness** [comment]
trigger publish-delegate

**Comment by abosii-harness** [comment]
trigger publish-delegate

**Comment by abosii-harness** [comment]
trigger publish-delegate

**Comment by abosii-harness** [comment]
trigger publish-delegate

**Comment by abosii-harness** [comment]
trigger publish-delegate

**Comment by abosii-harness** [comment]
trigger publish-delegate

**Comment by abosii-harness** [comment]
trigger publish-delegate

**Comment by abosii-harness** [comment]
trigger checkstylecodeformat, commonchecks

**Comment by abosii-harness** [comment]
trigger checkstylecodeformat, commonchecks

**Comment by abosii-harness** [comment]
trigger checkstylecodeformat

**Comment by abosii-harness** [comment]
trigger checkstylecodeformat, commonchecks

**Comment by abosii-harness** [comment]
trigger checkstylecodeformat, commonchecks

**Comment by abosii-harness** [comment]
trigger checkstylecodeformat, commonchecks

**Comment by abosii-harness** [comment]
trigger uts

**Comment by abosii-harness** [comment]
trigger publish-delegate

**Review by alebuh — DISMISSED** [review]
> Reviewed 27 of 27 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55409)* status: 27 file(s) to review. 1 of 2 approvals obtained. Waiting on team(s): del-team,k8s-team.

**Review by GoelPratyush99 — APPROVED** [review]

**Review by abhinavsingh1196 — DISMISSED** [review]
> Reviewed 27 of 27 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55409)* status: Waiting on team(s): del-team,k8s-team.

**Comment by abosii-harness** [comment]
trigger publish-delegate

**Comment by abosii-harness** [comment]
trigger checkstylecodeformat, commonchecks

**Comment by abosii-harness** [comment]
trigger publish-delegate

**Comment by abosii-harness** [comment]
trigger publish-delegate

**Comment by abosii-harness** [comment]
trigger publish-delegate, uts

**Comment by abosii-harness** [comment]
trigger checkstylecodeformat, commonchecks, uts

**Review by abhinavsingh1196 — COMMENTED** [review]
> Reviewed 4 of 5 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55409)* status: 5 file(s) to review. 0 of 2 approvals obtained. Waiting on team(s): del-team.

**Comment by abosii-harness** [comment]
trigger checkstylecodeformat, commonchecks, uts

File: 960-api-services/src/main/java/io/harness/helm/HelmCliExecutorFactory.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    return DEFAULT_MAX_CORE_POOL_SIZE;
+  }
+
+  private static int parseNumberValue(String number, int defaultValue) {
+    try {
+      return Integer.parseInt(number);
+    } catch (NumberFormatException e) {
+      log.warn("Invalid value provided, defaulting to value {}", defaultValue, e);
```
Comment (acasian-harness) [inline 960-api-services/src/main/java/io/harness/helm/HelmCliExecutorFactory.java:59]: shall we print the invalid value also for troubleshooting purposes?

**Review by acasian-harness — DISMISSED** [review]

**Comment by abosii-harness** [comment]
trigger checkstylecodeformat, commonchecks, uts

File: 960-api-services/src/main/java/io/harness/helm/HelmCliExecutorFactory.java (hunk 1 above)
Comment (abosii-harness) [reply 960-api-services/src/main/java/io/harness/helm/HelmCliExecutorFactory.java]: done

**Review by acasian-harness — APPROVED** [review]

**Review by alebuh — DISMISSED** [review]

**Comment by abosii-harness** [comment]
trigger checkstylecodeformat, commonchecks, uts

**Review by abhinavsingh1196 — APPROVED** [review]
> Reviewed 1 of 1 files at r4, 1 of 1 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55409)* status: 9 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained. Waiting on team(s): del-team.

**Comment by abosii-harness** [comment]
trigger checkstylecodeformat, commonchecks, uts

**Review by alebuh — APPROVED** [review]

**Review by markoharness — APPROVED** [review]
Really great and detailed description.

**Comment by abosii-harness** [comment]
trigger review

**Pull request merged**